pulpcore/plugin/serializers/content.py

```diff
@@ -74,8 +74,37 @@ class Meta:
         fields = NoArtifactContentSerializer.Meta.fields + UploadSerializerFieldsMixin.Meta.fields
 
 
+class _SingleArtifactContentSerializer(SingleArtifactContentSerializer):
+    """
+    Class that extends SingleArtifactContentSerializer to support re-uploading existing content
+    """
+
+    def create(self, validated_data):
+        """
+        Save a GenericContent unit or retrieve an already existing one.
+
+        This must be used inside a task that locks on the Artifact and if given, the repository.
+        """
+
+        content = self.retrieve(validated_data)
+        if content is not None:
+            content.touch()
+        else:
+            content = super().create(validated_data)
+
+        return content
+
+    def retrieve(self, validated_data):
+        """
+        Retrieve existing content unit if it exists, else return None.
+
+        This method is plugin-specific and therefore should be re-implemented inside each plugin.
```

Comment:
I wouldn't say it's mandatory.
```suggestion
        This method is plugin-specific and implementing it for a content type allows uploading already existing files.
```